File: LeetCode/python/lc_0414.py
"""
 * @file     lc_0414.py
 * @brief    https://leetcode-cn.com/problems/third-maximum-number/
 * @author   YongDu
 * @date     2021-11-01
"""


class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        n1st = n2nd = n3rd = -sys.maxsize - 1

        for n in nums:
            if n > n1st:
                n3rd, n2nd, n1st = n2nd, n1st, n
            elif n > n2nd and n < n1st:
                n3rd, n2nd = n2nd, n
            elif n > n3rd and n < n2nd:
                n3rd = n

        return n3rd if n3rd != -sys.maxsize - 1 else n1st

File: LeetCode/cpp/lc_0414.cpp
/**
 * @file     thirdMax.cpp
 * @brief
 * @author   DuYong
 * @date     2021-03-09
 */
class Solution {
  public:
    int thirdMax(vector<int>& nums) {
        std::set<int> numSet;
        for (int n : nums) {
            numSet.insert(n);
            if (numSet.size() > 3) {
                numSet.erase(numSet.begin());
            }
        }
        return numSet.size() == 3 ? *(numSet.begin()) : *(--numSet.end());
    }
};